File: Test/13_call_llm/test_call_llm.py
"""
CASSIA Test 13: call_llm Provider Testing
==========================================
Tests that call_llm correctly connects to different LLM providers.

This test verifies API connectivity for:
- OpenAI (gpt-4o)
- Anthropic (claude-sonnet-4-5)
- OpenRouter (google/gemini-2.5-flash)
- Custom/DeepSeek (deepseek-chat via https://api.deepseek.com)

Usage:
    python test_call_llm.py
"""

import os
import sys
import time
from pathlib import Path

# Add shared utilities to path
sys.path.insert(0, str(Path(__file__).parent.parent / "shared" / "python"))

from test_utils import (
    setup_cassia_imports,
    load_config,
    setup_api_keys,
    print_test_header,
    print_test_result,
    print_config_summary
)
from result_manager import (
    create_results_dir,
    save_test_metadata,
    save_test_results,
    create_test_metadata
)

# Setup CASSIA imports
setup_cassia_imports()

# Try to load local API keys file if it exists
try:
    import set_api_keys  # Load API keys from local file
    print("Loaded API keys from set_api_keys.py")
except ImportError:
    pass  # No local keys file, will use setup_api_keys() instead


def run_call_llm_test():
    """Test call_llm function with different providers."""
    print_test_header("13 - call_llm Provider Testing")

    # Load configuration
    config = load_config()
    print_config_summary(config)

    # Setup API keys
    setup_api_keys()

    # Import CASSIA call_llm function
    from llm_utils import call_llm

    # Create results directory
    results_dir = create_results_dir("13_call_llm")
    print(f"Results will be saved to: {results_dir}")

    # Test cases: (provider, model, env_var_name, display_name)
    test_cases = [
        ("openai", "gpt-4o", "OPENAI_API_KEY", "OpenAI"),
        ("anthropic", "claude-sonnet-4-5", "ANTHROPIC_API_KEY", "Anthropic"),
        ("openrouter", "google/gemini-2.5-flash", "OPENROUTER_API_KEY", "OpenRouter"),
        ("https://api.deepseek.com", "deepseek-chat", "CUSTOMIZED_API_KEY", "Custom (DeepSeek)"),
    ]

    # Simple test prompt
    test_prompt = "Reply with exactly one word: 'success'"

    start_time = time.time()
    errors = []
    test_results = {
        "provider_tests": []
    }

    passed_count = 0
    skipped_count = 0
    failed_count = 0
    total_tests = len(test_cases)

    for provider, model, env_var, display_name in test_cases:
        print("\n" + "=" * 50)
        print(f"Provider: {display_name}")
        print(f"  Model: {model}")
        print("=" * 50)

        test_result = {
            "provider": provider,
            "display_name": display_name,
            "model": model,
            "env_var": env_var,
            "status": "pending",
            "response_length": 0,
            "error": None
        }

        # Check if API key is set
        api_key = os.environ.get(env_var)
        if not api_key:
            print(f"  Status: [SKIP] {env_var} not set in environment")
            test_result["status"] = "skipped"
            skipped_count += 1
            test_results["provider_tests"].append(test_result)
            continue

        print(f"  Status: API key found")
        print(f"  Sending test prompt...")

        try:
            response = call_llm(
                prompt=test_prompt,
                provider=provider,
                model=model,
                temperature=0.3,
                max_tokens=50
            )

            if response and len(response) > 0:
                test_result["status"] = "passed"
                test_result["response_length"] = len(response)
                test_result["response_preview"] = response[:100] if len(response) > 100 else response
                print(f"  [OK] Response received ({len(response)} chars)")
                print(f"  Response: {response[:80]}{'...' if len(response) > 80 else ''}")
                passed_count += 1
            else:
                test_result["status"] = "failed"
                test_result["error"] = "Empty response received"
                print(f"  [X] FAILED: Empty response received")
                errors.append(f"{display_name}: Empty response")
                failed_count += 1

        except Exception as e:
            test_result["status"] = "failed"
            test_result["error"] = str(e)
            error_msg = str(e)
            # Truncate long error messages for display
            if len(error_msg) > 200:
                error_msg = error_msg[:200] + "..."
            print(f"  [X] ERROR: {error_msg}")
            errors.append(f"{display_name}: {str(e)[:100]}")
            failed_count += 1

        test_results["provider_tests"].append(test_result)

    duration = time.time() - start_time

    # Summary
    print("\n" + "=" * 50)
    print("SUMMARY")
    print("=" * 50)
    print(f"  Passed:  {passed_count}/{total_tests}")
    print(f"  Failed:  {failed_count}/{total_tests}")
    print(f"  Skipped: {skipped_count}/{total_tests}")
    print(f"  Duration: {duration:.2f}s")

    # Determine overall status
    # Test passes if all non-skipped tests passed
    tests_run = total_tests - skipped_count
    if tests_run == 0:
        status = "skipped"
        print("\n  Note: All tests skipped (no API keys configured)")
    elif passed_count == tests_run:
        status = "passed"
    else:
        status = "failed"

    # Save metadata and results
    metadata = create_test_metadata(
        test_name="call_llm_provider_test",
        config=config,
        duration_seconds=duration,
        status=status,
        clusters_tested=[],  # Not applicable for this test
        errors=errors
    )
    metadata["test_summary"] = {
        "passed": passed_count,
        "failed": failed_count,
        "skipped": skipped_count,
        "total": total_tests
    }
    save_test_metadata(results_dir, metadata)
    save_test_results(results_dir, test_results)

    # Print final result
    success = status == "passed"
    print_test_result(success, f"Duration: {duration:.2f}s")

    return success


if __name__ == "__main__":
    success = run_call_llm_test()
    sys.exit(0 if success else 1)

File: Test/13_call_llm/test_call_llm.R
# CASSIA Test 13: call_llm Provider Testing (R)
# ===============================================
# Tests that call_llm correctly connects to different LLM providers.
#
# This test verifies API connectivity for:
# - OpenAI (gpt-4o)
# - Anthropic (claude-sonnet-4-5)
# - OpenRouter (google/gemini-2.5-flash)
# - Custom/DeepSeek (deepseek-chat via https://api.deepseek.com)
#
# Usage:
#     Rscript test_call_llm.R

# Get script directory (works with Rscript and source())
get_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("--file=", "", file_arg), winslash = "/")))
  }
  for (i in sys.nframe():1) {
    if (!is.null(sys.frame(i)$ofile)) {
      return(dirname(normalizePath(sys.frame(i)$ofile, winslash = "/")))
    }
  }
  return(normalizePath(getwd(), winslash = "/"))
}
script_dir <- get_script_dir()

# Source shared utilities
source(file.path(script_dir, "..", "shared", "r", "test_utils.R"))
source(file.path(script_dir, "..", "shared", "r", "result_manager.R"))
source(file.path(script_dir, "..", "shared", "r", "logging_manager.R"))

run_call_llm_test <- function() {
  print_test_header("13 - call_llm Provider Testing (R)")

  # Load configuration
  config <- load_config()
  print_config_summary(config)

  # Setup API keys
  setup_api_keys()

  # Setup CASSIA R package (for Python interop via reticulate)
  tryCatch({
    setup_cassia()
    message("CASSIA R package loaded successfully")
  }, error = function(e) {
    message("Error loading CASSIA: ", e$message)
    return(FALSE)
  })

  # Create results directory
  results_dir <- create_results_dir("13_call_llm")

  start_logging(results_dir)
  log_msg("Results will be saved to:", results_dir)

  # Import Python call_llm function via reticulate
  cassia_module <- reticulate::import("CASSIA")
  call_llm <- cassia_module$call_llm

  # Test cases: list(provider, model, env_var_name, display_name)
  test_cases <- list(
    list(provider = "openai", model = "gpt-4o", env_var = "OPENAI_API_KEY", display_name = "OpenAI"),
    list(provider = "anthropic", model = "claude-sonnet-4-5", env_var = "ANTHROPIC_API_KEY", display_name = "Anthropic"),
    list(provider = "openrouter", model = "google/gemini-2.5-flash", env_var = "OPENROUTER_API_KEY", display_name = "OpenRouter"),
    list(provider = "https://api.deepseek.com", model = "deepseek-chat", env_var = "CUSTOMIZED_API_KEY", display_name = "Custom (DeepSeek)")
  )

  # Simple test prompt
  test_prompt <- "Reply with exactly one word: 'success'"

  start_time <- Sys.time()
  errors <- list()
  test_results <- list()

  passed_count <- 0
  skipped_count <- 0
  failed_count <- 0
  total_tests <- length(test_cases)

  for (tc in test_cases) {
    log_separator()
    log_msg("Provider:", tc$display_name)
    log_msg("  Model:", tc$model)
    log_separator()

    test_result <- list(
      provider = tc$provider,
      display_name = tc$display_name,
      model = tc$model,
      env_var = tc$env_var,
      status = "pending",
      response_length = 0,
      error = NULL
    )

    # Check if API key is set
    api_key <- Sys.getenv(tc$env_var, unset = "")
    if (api_key == "") {
      log_msg("  Status: [SKIP]", tc$env_var, "not set in environment")
      test_result$status <- "skipped"
      skipped_count <- skipped_count + 1
      test_results <- c(test_results, list(test_result))
      next
    }

    log_msg("  Status: API key found")
    log_msg("  Sending test prompt...")

    tryCatch({
      response <- call_llm(
        prompt = test_prompt,
        provider = tc$provider,
        model = tc$model,
        temperature = 0.3,
        max_tokens = as.integer(50)
      )

      if (!is.null(response) && nchar(response) > 0) {
        test_result$status <- "passed"
        test_result$response_length <- nchar(response)
        test_result$response_preview <- substr(response, 1, 100)
        log_msg("  [OK] Response received (", nchar(response), " chars)", sep = "")
        log_msg("  Response:", substr(response, 1, 80), if (nchar(response) > 80) "..." else "")
        passed_count <- passed_count + 1
      } else {
        test_result$status <- "failed"
        test_result$error <- "Empty response received"
        log_msg("  [X] FAILED: Empty response received")
        errors <- c(errors, paste0(tc$display_name, ": Empty response"))
        failed_count <- failed_count + 1
      }
    }, error = function(e) {
      test_result$status <<- "failed"
      test_result$error <<- e$message
      error_msg <- substr(e$message, 1, 200)
      log_msg("  [X] ERROR:", error_msg)
      errors <<- c(errors, paste0(tc$display_name, ": ", substr(e$message, 1, 100)))
      failed_count <<- failed_count + 1
    })

    test_results <- c(test_results, list(test_result))
  }

  duration <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))

  # Summary
  log_separator()
  log_msg("SUMMARY")
  log_separator()
  log_msg("  Passed: ", passed_count, "/", total_tests, "", sep = "")
  log_msg("  Failed: ", failed_count, "/", total_tests, "", sep = "")
  log_msg("  Skipped:", skipped_count, "/", total_tests, "", sep = "")
  log_msg("  Duration:", round(duration, 2), "s")

  # Determine overall status
  tests_run <- total_tests - skipped_count
  if (tests_run == 0) {
    status <- "skipped"
    log_msg("\n  Note: All tests skipped (no API keys configured)")
  } else if (passed_count == tests_run) {
    status <- "passed"
  } else {
    status <- "failed"
  }

  # Save metadata and results
  metadata <- create_test_metadata(
    test_name = "call_llm_provider_test",
    config = config,
    duration_seconds = duration,
    status = status,
    clusters_tested = list(),  # Not applicable for this test
    errors = errors
  )
  metadata$test_summary <- list(
    passed = passed_count,
    failed = failed_count,
    skipped = skipped_count,
    total = total_tests
  )
  save_test_metadata(results_dir, metadata)

  save_test_results(results_dir, list(
    provider_tests = test_results
  ))

  # Print final result
  success <- status == "passed"
  print_test_result(success, paste("Duration:", round(duration, 2), "s"))


  stop_logging()
  return(success)
}

# Run test
success <- run_call_llm_test()
quit(status = if (success) 0 else 1)
